### PR TITLE
WI #1389, detect if program name doesn't match file name

### LINKED ISSUE #1389: Detect if program name doesn't match file name
**Is your feature request related to a problem? Please describe.**
In our company, the program name must match the file name.
We should have a custom EI warning to enforce this.

TypeCobol code that you'd like to write.
File name `XXXF0YYY`:
```cobol
       IDENTIFICATION DIVISION .
      *No warning here
       PROGRAM-ID. XXXF0YYY.
      
       END PROGRAM XXXF0YYY.
```

File name `XXXF0YYY`:
```cobol
       IDENTIFICATION DIVISION .
      *Warning here because it doesn't match file name.
       PROGRAM-ID. AAAF0YYY.
      
       END PROGRAM AAAF0YYY.
```

### PR DESCRIPTION
**Pull request by LANOYDO**
Fixes #1389

### CODE REVIEW

File: TypeCobol.LanguageServer/Workspace.cs [hunk 1]
```diff
@@ -422,6 +429,7 @@ public void DidChangeConfigurationParams(IEnumerable<string> arguments)
 
 #if EUROINFO_RULES
             typeCobolOptions.AutoRemarksEnable = TypeCobolConfiguration.AutoRemarks;
+            typeCobolOptions.UseCheckProgramName = UseCheckProgramName;
```
Comment (maxime645) [inline TypeCobol.LanguageServer/Workspace.cs]: The option can be used in a normal environment. The value UseCheckProgramName should be set in the constructor of TypeCobolOptions.
Use the value stored in TypeCobolConfiguration that is retrieved from the command line. If the configuration of the client change the new value won't be taken into account.

File: TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs [hunk 2]
```diff
@@ -987,6 +989,14 @@ public static void OnNode(Program node)
                     "\"END PROGRAM\" is missing.", MessageCode.Warning);
             }
 
+            if (node.IsMainProgram && compilerOptions.CheckProgramName)
+            {
+                string shortFilename = Path.GetFileNameWithoutExtension(node.CodeElement.TokenSource.SourceName);
+                if (!node.Name.Equals(shortFilename, StringComparison.InvariantCultureIgnoreCase))
```
Comment (smedilol) [inline TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: For performance reason, please use StringComparison.OrdinalIgnoreCase

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 3]
```diff
@@ -15,6 +15,12 @@ namespace TypeCobol.Compiler.Diagnostics
 {
     public class CrossCompleteChecker : AbstractAstVisitor
     {
+        public CrossCompleteChecker(string sourceName = "")
+        {
+            SourceName = sourceName;
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: done

File: TypeCobol/Compiler/Directives/TypeCobolOptions.cs [hunk 4]
```diff
@@ -37,19 +37,42 @@ public bool UseEuroInformationLegacyReplacingSyntax
...
+        private bool _useCheckProgramName = true;
 #else
         private bool _useEuroInformationLegacyReplacingSyntax;
+        private bool _checkProgramName = false;
+        private bool _useCheckProgramName = false;
 #endif
+        /// <summary>
+        /// Check if program name matches the file name.
+        /// </summary>
+        public bool CheckProgramName
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/Directives/TypeCobolOptions.cs]: done

File: TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs (hunk 2 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: done

File: TypeCobol.Test/TestCollection.cs [hunk 5]
```diff
@@ -310,5 +310,42 @@ public void CheckDocumentation()
...
+
+            string directory = PlatformUtils.GetPathForProjectFile("Parser" + Path.DirectorySeparatorChar + "ToBeIndividuallyExecuted");
+            var folderTester = new FolderTester(directory, directory, directory, extensions, compilerExtensions);
+            try
+            {
+                folderTester.Test(false, false, false, true);
+            }
+            catch (Exception ex)
+            {
+                errors.Add(ex);
```
Comment (fm-117) [inline TypeCobol.Test/TestCollection.cs]: I don't get the logic here. As soon as there is one exception, you'll leave the folderTester.Test method so there can be only one exception. The `errors` list is useless.
I think you'll need to iterate over the files to have true individual testing but there should be no need for that.

File: TypeCobol.Test/Utils/Comparator.cs [hunk 6]
```diff
@@ -26,14 +27,16 @@ public TestUnit(FilesComparator comparator1, bool debug = false)
...
+        public void Init(string[] extensions = null, bool autoRemarks = false, bool AntlrProfiler = false, bool checkProgramName = false)
         {
             DirectoryInfo localDirectory = new DirectoryInfo(Path.GetDirectoryName( Comparator?.paths?.SamplePath));
             DocumentFormat format = Comparator?.GetSampleFormat();
             TypeCobolOptions options = new TypeCobolOptions();
 #if EUROINFO_RULES
-            options.AutoRemarksEnable = autoRemarks;
+            options.AutoRemarksEnable = autoRemarks;    
 #endif
+            if (!checkProgramName) options.CheckProgramName = new TypeCobolCheckOption(false);
```
Comment (fm-117) [inline TypeCobol.Test/Utils/Comparator.cs]: `options.CheckProgramName = new TypeCobolCheckOption(checkProgramName)` ?

File: TypeCobol/Parser.cs [hunk 7]
```diff
@@ -149,12 +150,14 @@ public IEnumerable<CodeElement> CodeElements
...
 
 
-		public static Parser Parse(string path, DocumentFormat format, bool autoRemarks = false, IList<string> copies = null) {
+		public static Parser Parse(string path, DocumentFormat format, bool autoRemarks = false, IList<string> copies = null, bool checkProgramName = true) {
 			var parser = new Parser();
             var typeCobolOption = new TypeCobolOptions() { ExecToStep = ExecutionStep.Generate };
 #if EUROINFO_RULES
 		    typeCobolOption.AutoRemarksEnable = autoRemarks;
 #endif
+            if (!checkProgramName) typeCobolOption.CheckProgramName = new TypeCobolCheckOption(false);
```
Comment (fm-117) [inline TypeCobol/Parser.cs]: Same remark: equivalent to `typeCobolOption.CheckProgramName = new TypeCobolCheckOption(checkProgramName);`.

File: TypeCobol/Tools/Options-Config/TypeCobolConfiguration.cs [hunk 8]
```diff
@@ -14,7 +16,7 @@ public class TypeCobolConfiguration
     {
         public string CommandLine { get; set; }
         public DocumentFormat Format = DocumentFormat.RDZReferenceFormat;
-        public bool AutoRemarks;
+        public bool AutoRemarks;
```
Comment (fm-117) [inline TypeCobol/Tools/Options-Config/TypeCobolConfiguration.cs]: Unnecessary spacing.

**Comment by fm-117** [comment]
This is a global remark about the configuration system. Currently adding new capabilities via option introduce a lot of duplicate code (TypeCobolServer.cs, TypeCobolServerHost.cs, TypeCobolOptions.cs, TypeCobolConfiguration.cs...).
Maybe we should consider in a new issue improving or simply cleaning up the configuration system.

**Comment by smedilol** [comment]
After discussion, ok to set this diagnostic to Ignore by default

**Comment by smedilol** [comment]
Our RDZ client need to manipulate file without changing its name before we can use this diagnostic.

**Comment by fm-117** [comment]
@smedilol this is checked in our private quality analyzer, can we close it ?

**Comment by smedilol** [comment]
Done with our private Quality analyzer.

**Pull request closed without merging**